Bugfix: 3075

Xbar opened the pull request:
1. Fix LaRank freeze Issue: https://github.com/shogun-toolbox/shogun/issues/3075
2. Minor style changes to LaRank unit test

vigsterkr commented on src/shogun/multiclass/LaRank.h:
```diff
@@ -485,6 +485,8 @@ namespace shogun
 			/// progess output
 			int32_t step;
+                        
+                        static const int32_t MAX_ITERATION = 1000;
```
mmm why a static, why not a settable property of LaRank?

karlnapf replied:
definitely

karlnapf replied:
Also, something more well informed. How many iterations does this usually take until convergence? What does the literature say here?

vigsterkr commented on tests/unit/multiclass/LaRank_unittest.cc:
```diff
@@ -6,56 +6,54 @@
-	index_t num_feat=3;
-	index_t num_class=num_feat; // to make data easy
-	float64_t distance=15;
+	index_t num_vec = 10;
```
please refrain from whitespace refactorings (see this whole unittest patch)... it just makes git bisecting more complicated. just make sure that your patches are following the codestyle requirements

karlnapf replied:
the best thing is to do a first patch with the bugfix, and a second with the whitespace changes

karlnapf replied:
Currently it is not clear what actually changed

karlnapf commented on src/shogun/multiclass/LaRank.cpp:
```diff
@@ -653,16 +653,20 @@ bool CLaRank::train_machine(CFeatures* data)
 	SG_DONE()
 
+        if (n_it >= MAX_ITERATION && gap > get_C())
+            SG_WARNING("LaRank exceeded max iteration but not converged.")
```
La Rank did not converge after the maximum number of %d iterations

karlnapf commented:
Great! Thanks for the patch.
Can you address our minor comments? I think it can be merged then....travis is fine

Xbar commented:
Changed max_iteration to be an option
Reverted most changes to the unit test

Max iteration is hard to set, because morbid cases can always be made that will take very long to converge. In practice, if iteration exceeds a fairly large number, it may reflect some problems in the training set, or how SVM is initialized.

vigsterkr commented on src/shogun/multiclass/LaRank.h:
```diff
@@ -378,6 +378,17 @@ namespace shogun
 			 * @return tau
 			 */
 			float64_t get_tau() { return tau; };
+                        /** set max iterations
```
indentation on this is wrong

vigsterkr commented on src/shogun/multiclass/LaRank.h:
```diff
@@ -485,6 +496,8 @@ namespace shogun
 			/// progess output
 			int32_t step;
+                        
+                        int32_t max_iteration;
```
same here

vigsterkr commented:
@Xbar what all these unrelated doxygen changes? i mean it's great but it shoudl be part of separate PR, not about fixing LaRank

Xbar commented:
I was trying to do a rebase, but I didn't do it the right way. Somehow it merged other commits from other people.

Xbar commented:
@vigsterkr Could you close this PR?
Sorry but I really made a mess in my forked repo after an improper rebase.
I deleted my repo and created a new PR.
